src/overlays/effects/ovl_Effect_Ss_D_Fire/z_eff_ss_d_fire.c

```diff
@@ -1,31 +1,114 @@
 /*
  * File: z_eff_ss_d_fire.c
  * Overlay: ovl_Effect_Ss_D_Fire
- * Description:
+ * Description: Dodongo Fire
  */
 
 #include "z_eff_ss_d_fire.h"
+#include "objects/object_dodongo/object_dodongo.h"
+
+#define rScale regs[0]
+#define rTexIndex regs[1]
+#define rAlpha regs[5]
+#define rAlphaStep regs[6]
+#define rFadeDelay regs[7]
+#define rScaleStep regs[8]
+#define rObjectIndex regs[9]
 
 #define PARAMS ((EffectSsDFireInitParams*)initParamsx)
 
-s32 EffectSsDFire_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx);
+u32 EffectSsDFire_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx);
 void EffectSsDFire_Update(PlayState* play, u32 index, EffectSs* this);
 void EffectSsDFire_Draw(PlayState* play, u32 index, EffectSs* this);
 
-#if 0
 const EffectSsInit Effect_Ss_D_Fire_InitVars = {
     EFFECT_SS_D_FIRE,
     EffectSsDFire_Init,
 };
 
-#endif
+static TexturePtr sFireTextures[] = { gDodongoFire0Tex, gDodongoFire1Tex, gDodongoFire2Tex, gDodongoFire3Tex };
+
+s32 EffectSsDFire_CheckForObject(EffectSs* this, PlayState* play) {
+    if (((this->rObjectIndex = Object_GetIndex(&play->objectCtx, OBJECT_DODONGO)) < 0) ||
+        !Object_IsLoaded(&play->objectCtx, this->rObjectIndex)) {
+        this->life = -1;
+        this->draw = NULL;
+        return false;
+    }
+    return true;
+}
+
+u32 EffectSsDFire_Init(PlayState* play, u32 index, EffectSs* this, void* initParamsx) {
+    EffectSsDFireInitParams* initParams = PARAMS;
+
+    if (EffectSsDFire_CheckForObject(this, play)) {
+        Math_Vec3f_Copy(&this->pos, &initParams->pos);
+        Math_Vec3f_Copy(&this->velocity, &initParams->velocity);
+        Math_Vec3f_Copy(&this->accel, &initParams->accel);
+        this->gfx = gDodongoFireDL;
+        this->life = initParams->life;
+        this->rScale = initParams->scale;
+        this->rScaleStep = initParams->scaleStep;
+        this->draw = EffectSsDFire_Draw;
+        this->update = EffectSsDFire_Update;
+        this->rTexIndex = (play->state.frames % 4) ^ 3;
```

Comment:
Curious `^ 3`

Reply:
Yeah, this smells like an optimisation. Did it have to be %?

Reply:
Nah `& 3` also works so you would have `(x & 3) ^ 3` which does seem like an optimization or something, but I can't quite put my finger on it.

Reply:
I think unless we figure it out I'd rather have `(x & 3) ^ 3` since at least then it's clearer what it's doing.

Reply:
Looking at this again, I think the `^ 3` is just toggling the direction, so `(x & 3)` would be going `0 -> 1 -> 2 -> 3 -> 0 -> 1 -> 2 -> 3` every frame, but adding the `^ 3` makes it go downwards as frames advance, so `3 -> 2 -> 1 -> 0 -> 3 -> 2 -> 1 -> 0`.

Reply:
Yeah that is what it is doing (though only on initialization so still cycles forwards in update).

Reply:
Guessing that `3 - ` doesn't match, though.

Reply:
No it does not.

Reply:
I'd probably leave a note on what it's doing, then.